[Bulk Update Orders] Enable multiple selection in Order List: Part II

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #13130, #13150

please do not merge until target is `trunk`
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds various UI-related updates to Orders list screen, related to the bulk update feature. Those updates are:

1. Adds a checkbox showing when an Order is selected in multi select mode, similar to how it works in Products screen.
2. Hides bottom nav bar when multi selection mode is active, similar to behavior on Products.
3. Disables filter button when selection mode is active, similar to behavior on Products.
4. Disables pull to refresh when selection mode is active, similar to behavior on Products.
5. Hides Create Order FAB when selection mode is active, similar to behavior on Products.
6. Disables navigation to Order details on edge cases, e.g: when using Talkback


### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Build app, go to Orders,
2. Long press an Order to enable multiple selection, ensure it shows the checkbox image next to it.
3. Play around with selecting/deselecting other Orders.
4. Ensure that bottom nav bar is hidden when multiple selection is active.
7. Ensure "Filter" button is inactive when multiple selection is active.
8. Ensure pull to refresh can't be done when multiple selection is active.
9. Ensure Create Order FAB is hidden when multiple selection is active.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Test this both in phone and tablet (2 panel) mode, the test items should apply to both.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->

I've tested the steps above both on phone and tablet simulator, API 32 and 34.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/user-attachments/assets/a5da90c6-9c49-405e-8362-6ee3114d084b



- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hafizrahman committed:
Update order list item layout to add a checkbox for selection state.

hafizrahman committed:
Update UI so that the checkbox appears when item is selected, and remove the bg color.

hafizrahman committed:
Detekt fix

hafizrahman committed:
Merge branch 'task/13130-enable-order-multi-select' into task/13130-order-multi-select-ii

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Add logic to hide bottom nav bar also during multi selection mode.

This requires some refactoring on how isSearching is stored, since hiding the bottom bar has to check for both searching and selecting state.

hafizrahman committed:
Disable order filter button when bulk selection is active.

hafizrahman committed:
Disable pull to refresh option when bulk selection is active.

hafizrahman committed:
Merge branch 'task/13130-enable-order-multi-select' into task/13130-order-multi-select-ii

irfano reviewed (changes requested):
Before starting to review code, I want to list issues I found while testing:

- In selection mode, when tapped the create order button, the title remains "`order selected"
<details>
  <summary>Video</summary> 

[new-order.webm](https://github.com/user-attachments/assets/e93d7d45-8b8e-4ebf-8d1b-e0a8522fc70d)
</details>

- In selection mode, when changing the orientation, the title remains "`order selected".
<details>
  <summary>Video</summary> 

[orientation.webm](https://github.com/user-attachments/assets/030b185e-1dd8-4680-b0dc-cd23c700a572)
</details>

- When Talkback is enabled, we can select the order with double tap but holding the second tap. In selection mode, I am able to open an order. I think I shouldn't be able to do it, as I can't do with the product list.
<details>
  <summary>Video</summary> 

[talkback.webm](https://github.com/user-attachments/assets/ab6a6781-b9a6-4f06-ae1e-8125abe57a27)
</details>

hafizrahman committed:
Hide add order FAB when multi selection is active

hafizrahman committed:
Disable navigation into Order Detail on edge cases like Talkback usage.

Similar to implementation in product https://github.com/woocommerce/woocommerce-android/pull/8160

hafizrahman commented:
Thanks for the check @irfano !

> In selection mode, when tapped the create order button, the title remains "`order selected"

Like in Product, the create order button should be hidden in selection mode. Added in https://github.com/woocommerce/woocommerce-android/pull/13151/commits/2976ce6ee027a20d27d279c60c514c96744e2bd1

> In selection mode, when changing the orientation, the title remains "`order selected".

I haven't worked on persisting selection state during configuration change yet as it's a bit non-trivial. If you see there, the selected products are lost, too. I will work on this as a separate PR, the task is already tracked in https://github.com/woocommerce/woocommerce-android/issues/13130

> When Talkback is enabled, we can select the order with double tap but holding the second tap. In selection mode, I am able to open an order. I think I shouldn't be able to do it, as I can't do with the product list.

I added a commit on https://github.com/woocommerce/woocommerce-android/pull/13151/commits/59aa27e8f1e2e6d052f4fc9de48a729bb30d3c5a to deal with this, it's a similar fix as Products in https://github.com/woocommerce/woocommerce-android/pull/7971

hafizrahman committed:
Merge branch 'trunk' into task/13130-order-multi-select-ii

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListFragment.kt:
```diff
@@ -726,6 +728,36 @@ class OrderListFragment :
+                uiMessageResolver.anchorViewId = null
+                binding.createOrderButton.hide()
+            }
+        }
```
np: I know this is copied from Products, but I see no point in using `when` this way when the options are only `true` and `else`. I suggest using `if` here:

```
if(show) {
     uiMessageResolver.anchorViewId = binding.createOrderButton.id
     binding.createOrderButton.show()
} else {
     uiMessageResolver.anchorViewId = null
     binding.createOrderButton.hide()
}
```

hafizrahman replied:
good point https://github.com/woocommerce/woocommerce-android/pull/13151/commits/dc6169466be0eb65ca074f4fcb7f7e9dddd50870

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListFragment.kt:
```diff
@@ -826,13 +858,29 @@ class OrderListFragment :
+        if (viewModel.isSelecting()) {
+            tracker?.let { selectionTracker ->
+                if (selectionTracker.isSelected(orderId)) selectionTracker.deselect(orderId)
+                else selectionTracker.select(orderId)
```
Detekt flagged the code style here and is showing a warning.

hafizrahman replied:
Thanks! https://github.com/woocommerce/woocommerce-android/pull/13151/commits/93331462e116163436f02adb61e94a48a9d4e2da

irfano commented on WooCommerce/src/main/res/layout/order_list_item.xml:
```diff
@@ -52,27 +75,32 @@
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintTop_toBottomOf="@id/orderName" />
+
+        <Space
+            android:id="@+id/spacer"
+            android:layout_width="match_parent"
+            android:layout_height="@dimen/major_100"
+            app:layout_constraintTop_toBottomOf="@+id/orderTags" />
```
I’m curious why you decided to add this Space.

hafizrahman replied:
When I added the changes, the divider somehow did not respect vertical spacing, so I resorted to adding spacing inbetween. In the next refactor https://github.com/woocommerce/woocommerce-android/pull/13151/commits/bb014fda0388a7f701e31d90148abd4567fb997d this is no longer needed, though.

irfano commented on WooCommerce/src/main/res/layout/order_list_item.xml:
```diff
@@ -52,27 +75,32 @@
+            android:layout_marginEnd="@dimen/major_100"
             app:layout_constraintEnd_toEndOf="parent"
             app:layout_constraintTop_toBottomOf="@+id/orderDate"
             tools:text="$100"/>
 
         <com.woocommerce.android.widgets.FlowLayout
             android:id="@+id/orderTags"
             style="@style/Woo.FlowLayout"
-            android:layout_width="wrap_content"
+            android:layout_width="0dp"
             android:layout_height="wrap_content"
-            app:layout_constraintBottom_toTopOf="@+id/divider"
-            app:layout_constraintHorizontal_bias="0.0"
-            app:layout_constraintStart_toStartOf="parent"
-            app:layout_constraintTop_toBottomOf="@+id/orderName"/>
+            app:layout_constraintStart_toEndOf="@id/start_barrier"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintTop_toBottomOf="@id/orderName" />
+
+        <Space
+            android:id="@+id/spacer"
+            android:layout_width="match_parent"
+            android:layout_height="@dimen/major_100"
+            app:layout_constraintTop_toBottomOf="@+id/orderTags" />
 
         <View
             style="@style/Woo.Divider"
             android:id="@+id/divider"
             android:layout_width="0dp"
-            app:layout_constraintTop_toBottomOf="@+id/orderTags"
-            app:layout_constraintBottom_toBottomOf="parent"
-            app:layout_constraintStart_toStartOf="@+id/orderNum"
+            app:layout_constraintTop_toBottomOf="@id/spacer"
+            app:layout_constraintStart_toStartOf="parent"
```
Divider was not starting from the start, but now it does. Did you make this change intentionally?

**before/after**
![image](https://github.com/user-attachments/assets/452e2751-a31a-43d0-9d77-87ab16a5720b)

hafizrahman replied:
This was because in the original state, when the checkbox is shown, the divider gets pushed to the right (because it follows the start of the order number) and becomes pretty weird:

<img src="https://github.com/user-attachments/assets/de5af09f-e295-4b3c-848a-a55b872c2b24" width="50%">

The divider on Products also look like that, but it has the benefit of always having product thumbnail on the left, so the divider line is always pushed to the right whether checked or not:

<img src="https://github.com/user-attachments/assets/98e5aadc-c1e8-4c12-8425-ff52909d295a" width="50%">

So, I made the start constraint to match the parent instead as it's the element that's always available as anchor.

irfano commented on WooCommerce/src/main/res/layout/order_list_item.xml:
```diff
@@ -13,12 +13,35 @@
+            android:visibility="gone"
+            tools:visibility="visible"
+            app:layout_constraintTop_toTopOf="parent"
+            app:layout_constraintStart_toStartOf="parent"/>
```
Do we have a Figma design for this checkbox? I think it could look better if it was centered vertically. WDYT?

![image](https://github.com/user-attachments/assets/d09734ed-98b9-4742-9589-e161f8fb93dd)

hafizrahman replied:
Added in https://github.com/woocommerce/woocommerce-android/pull/13151/commits/bb014fda0388a7f701e31d90148abd4567fb997d

hafizrahman committed:
simplify if logic

hafizrahman committed:
Detekt fix

hafizrahman committed:
Various refactor to match the design on Products screen.

hafizrahman commented:
@irfano 

Thanks for your design reviews!

> Do we have a Figma design for this checkbox?

The constraint for this project is that we don't have any specific design, we just want to follow Products list. As I worked on this, though, I noticed that there are differences between the two screens, so replicating it exactly is not possible. For example, Products screen always has a thumbnail area, which gets replaced by the checkbox in selection mode. Orders screen doesn't have this, so the checkbox will disrupt the layout more in comparison. This affects smaller things like divider lines, etc, which you noticed in the review.

As there's no specific design direction, I just opt to work on something that makes sense visually. I added a refactor commit https://github.com/woocommerce/woocommerce-android/pull/13151/commits/bb014fda0388a7f701e31d90148abd4567fb997d that incorporates your feedback, while also tries to make it more similar to Products. These includes:

- Checkbox size now matches the size in Products
- Checkbox horizontal spacing now matches the size in Products
- Checkbox is now vertically centered
- Order item divider now retains the original horizontal spacing at the start
- Checkbox now has no transparency because it makes no sense when there's nothing below it, unlike in Products

This is how it looks like now:

<img src="https://github.com/user-attachments/assets/76836e81-fd0d-40b9-9c9d-8cd25f105d1d" width="42%">


I think this would be a good starting point for further design discussion, so if you have more feedbacks, I'd love to hear it. Since we don't have specific design to follow, I think we can aim for something that is essentially usable and not confusing to merchants.

irfano commented on WooCommerce/src/main/res/layout/order_list_item.xml:
```diff
@@ -52,27 +74,29 @@
             style="@style/Woo.Divider"
             android:id="@+id/divider"
             android:layout_width="0dp"
+            android:layout_marginStart="@dimen/minor_100"
```
The `marginStart` was previously coming from `orderNum`, and its value is `major_100`. I'd suggest keeping it unchanged.

```suggestion
            android:layout_marginStart="@dimen/major_100"
```

irfano commented on WooCommerce/src/main/res/layout/order_list_item.xml:
```diff
@@ -28,21 +50,21 @@
             android:layout_height="wrap_content"
             android:layout_marginTop="@dimen/minor_50"
-            app:layout_constraintStart_toStartOf="parent"
+            android:layout_marginStart="@dimen/major_100"
```
np: We don't need to set `...marginStart="@dimen/major_100"` for `orderNum` and `orderDate` because their initial `marginStart` is already the same and coming from [Woo.TextView](https://github.com/woocommerce/woocommerce-android/blob/d055dc3c5e28cc7fabb88db71850651881887090/WooCommerce/src/main/res/values/styles_base.xml#L208-L213). Just wanted to point this out, feel free to ignore.

hafizrahman replied:
Good catch! https://github.com/woocommerce/woocommerce-android/pull/13151/commits/96c8072a8695d6db0ebaf49a4f33bc1dc6150266

irfano reviewed (approved):
Thanks for explaining everything and sharing screenshots for edge cases, @hafizrahman! LGTM! I've just added a minor feedback. I believe we're good to merge after that. I’m pre-approving it. Great work! 🚀

hafizrahman committed:
Keep margin consistent.

Co-authored-by: Irfan Ömür <irfan.omur@automattic.com>

hafizrahman committed:
Remove already-set marginStart
